gui-lib/framework/private/text-port.rkt

```diff
@@ -538,50 +544,159 @@
-                 [else (values (car fst) (cdr fst))]))
-           
-             (define inserted-count
-               (if (is-a? str/snp snip%)
-                   (send str/snp get-count)
-                   (string-length str/snp)))
-             (define old-insertion-point insertion-point)
-             (set! insertion-point (+ insertion-point inserted-count))
-             (set! unread-start-point (+ unread-start-point inserted-count))
-           
-             (insert (if (is-a? str/snp snip%)
-                         (let ([s (send str/snp copy)])
-                           (if (is-a? s snip%)
-                               s
-                               (new snip%)))
-                         str/snp)
-                     old-insertion-point
-                     old-insertion-point
-                     #t)
-           
-             ;; the idea here is that if you made a string snip, you
-             ;; could have made a string and gotten the style, so you
-             ;; must intend to have your own style.
-             (unless (is-a? str/snp string-snip%)
-               (change-style style old-insertion-point insertion-point))
+                 ((string? markup)
+                  (send text insert markup))
+                 ((empty-markup? markup) (void))
+                 ((horizontal-markup? markup)
+                  (for-each (lambda (markup)
+                              (insert-markup markup text style #t))
+                            (horizontal-markup-markups markup)))
+                 ((vertical-markup? markup)
+                  (if inline?
+                      (send text insert (markup->snip markup style #f))
+                      (for-each/between (lambda (markup)
+                                          (insert-markup markup text style #f))
+                                        (lambda () (send text insert #\newline))
+                                        (vertical-markup-markups markup))))
+                 ((srcloc-markup? markup)
+                  (insert-srcloc-markup markup text style))
+                 ((framed-markup? markup)
+                  (send text insert (markup->snip (framed-markup-markup markup) style #t)))
+                 ((image-markup? markup)
+                  (send text insert (image-markup->snip markup style)))))
+             
+             (define (image-markup->snip markup style)
+               (let ((data (image-markup-data markup)))
+                 (cond
+                   ((is-a? data snip%) data)
+                   ((snip-special? data)
+                    (define snip (snip-special->snip data))
+                    (if (exn:fail? snip)
+                        (markup->snip (image-markup-alt-markup markup) style #f)
+                        snip))
+                   (else
+                    (markup->snip (image-markup-alt-markup markup) style #f)))))
+             
+             (define (record-dc-datum->bitmap datum width height)
```

Comment:
Why not just draw into the dc instead of creating a bitmap?

Reply:
Which dc? Don't I need a bitmap to create a `bitmap-dc%`?

Reply:
I'm saying that you could save the recorded-datum and then later, just use it to draw.

I suppose this means that we'd need a record-dc-snip% (which isn't entirely trivial, I am now realizing but perhaps not that bad).